### PR TITLE
Add snapshot subcommand that creates a ledger snapshot from an archive

### LINKED ISSUE #383: Add command for creating ledger snapshot
### What

Add a command for creating a ledger snapshot JSON file.

The ledger snapshot JSON file is the file format that the soroban-cli uses when in sandbox/local mode. The file format is also supported by the soroban-sdk in its test framework, and contract developers can load a ledger snapshot in a test by calling [`Env::from_snapshot_file(path)`](https://docs.rs/soroban-sdk/latest/soroban_sdk/struct.Env.html#method.from_snapshot_file). Contract developers can do this to get a ledger that has contracts and contract data in it, and write integration tests that execute their own contract alongside those other contracts.

### Why

To help developers create realistic integration test data for their contract tests, by helping them mirror pubnet/testnet data into their contract tests.

### Examples

Generate ledger.json containing all ledger entries of ledger 5.
```
$ soroban lab snapshot create --ledger 5
```

Generate ledger.json containing all ledger entries of ledger 5 that are owned by two contracts.
```
$ soroban lab snapshot create --ledger 5 --contract C1... --contract C2...
```

Generate ledger.json containing all ledger entries of ledger 5 that are owned by two contracts and one account.
```
$ soroban lab snapshot create --ledger 5 --contract C1... --contract C2... --account G1...
```

Generate ledger.json containing ledger entries of ledger 5 that have the given ledger keys.
```
$ soroban lab snapshot create --ledger 5 --ledger-key ...
```

### PR DESCRIPTION
**Pull request by leighmcculloch**
### What
Add snapshot subcommand that creates a ledger snapshot from an archive.

### Why
To make it possible to easily test against mainnet state. 

### Try it out

Install stellar-cli:
```
cargo install \
  --locked \
  --git https://github.com/stellar/stellar-cli \
  --branch snap \
  stellar-cli
```

Snapshot ledger:
```
stellar snapshot \
  --output json
  --address CCW67TSZV3SSS2HXMBQ5JFGCKJNXKZM7UQUWUZPUTHXSTZLEO7SJMI75 \
  --out snapshot.json
```

Write a test importing it:
```rust
#![cfg(test)]
use soroban_sdk::{token::TokenClient, Address, Env, String};

#[test]
fn test() {
    let e = Env::from_ledger_snapshot_file("snapshot.json");
    let contract = TokenClient::new(
        &e,
        &Address::from_string(&String::from_str(
            &e,
            "CCW67TSZV3SSS2HXMBQ5JFGCKJNXKZM7UQUWUZPUTHXSTZLEO7SJMI75",
        )),
    );
    let wallet = Address::from_string(&String::from_str(
        &e,
        "CAZ6W4WHVGQBGURYTUOLCUOOHW6VQGAAPSPCD72VEDZMBBPY7H43AYEC",
    ));
    assert_eq!(contract.balance(&wallet), 553804322934);
}
```

Close #383

### CODE REVIEW

**Comment by leighmcculloch** [comment]
Leaving this as a note for myself, that when searching for entries we can expect that bucket lists sort their entries according to [LedgerCmp.h](https://github.com/stellar/stellar-core/blob/master/src/bucket/LedgerCmp.h). Ref: https://stellarfoundation.slack.com/archives/C9H5JTW1J/p1718294323240629?thread_ts=1718292396.365429&cid=C9H5JTW1J (from @SirTyson)

**Comment by leighmcculloch** [comment]
```
❯ stellar snapshot --network mainnet --output json --contract-id CCOQM6S7ICIUWA225O5PSJWUBEMXGFSSW2PQFO6FP4DQEKMS5DASRGRR
🌎 Downloading history https://history.stellar.org/prd/core-live/core_live_001/.well-known/stellar-history.json
ℹ️  Ledger: 52827455
ℹ️  Network Passphrase: Public Global Stellar Network ; September 2015
ℹ️  Network ID: 7ac33997544e3175d266bd022439b22cdb16508c01163f26e5cb2a3e1045a979
🪣  Downloading bucket 0 5b071ae54a26e2295e7138c862d01ace6b743ce3acc9e7d6b2b1195426111474 (61.4 KB)
🪣  Downloading bucket 1 503e82f7895a420262dc8ef5e386b1e1cb3155ddee82179ea25567821a0dd8f8 (117.4 KB)
🪣  Downloading bucket 2 85deb9141ef1eaf725ea4ffb6c4278e75d5a0bdec38e655ecd1ff394441afc43 (355.4 KB)
🪣  Downloading bucket 3 0466a1292a7272e84bc8c1dd6c92de0d413ad5115d27bef60524d250cc627c94 (373.2 KB)
🪣  Downloading bucket 4 474f7af3cebcdb07de10357bb2fe21f7a726b4676a07b1927725ce0aaf263b45 (872.1 KB)
🪣  Downloading bucket 5 cce39f76191873624b6c3cacef7c0ea84aa28cc61eed91837c3008515c8ea1d5 (900.1 KB)
🪣  Downloading bucket 6 14de1cf5ddfa70ef8f193eff96de005650b335303199eceb489627ff7f1998ab (1.4 MB)
🪣  Downloading bucket 7 d374770c4c495225e235ac89d90586d58910bcf9c5e5386038b950d82294b796 (2.5 MB)
🪣  Downloading bucket 8 7cf48505829612746c83b40c74a79fb846456eed78bc3744f9728859a9660d5b (3.0 MB)
🪣  Downloading bucket 9 344f47a8f98aca94d7fc37671e676bd4d519bc6dd9aba9214edfe1ed86d39649 (5.1 MB)
🪣  Downloading bucket 10 fe44851e5f29e6a657e76479c46976f2e73c953dc899bd6d3f2b68b94c5b20ca (30.1 MB)
🪣  Downloading bucket 11 8d0a85ef79d8f26b69d2ef539a2fd0cfe8438990f38ae35e73c1f27f943177f2 (38.0 MB)
🪣  Downloading bucket 12 07bac08bd614a20ea079e888b0bfed0bfae7a38bbcefb1b5b09135f791cb006f (65.6 MB)
🪣  Downloading bucket 13 c258315f26e489954556bd61b0cae4f894d982f63130bfc3697d879165b1ccfd (68.9 MB)
🪣  Downloading bucket 18 a146de34263062d255a2373c2a0152721bf7e5dece0f12106b065d674c7e2c3b (468.5 MB)
ℹ️  Searching for 0 accounts, 1 contracts, 0 wasms
🔎 Looking in bucket 0 5b071ae54a26e2295e7138c862d01ace6b743ce3acc9e7d6b2b1195426111474
🔎 Looking in bucket 1 503e82f7895a420262dc8ef5e386b1e1cb3155ddee82179ea25567821a0dd8f8
🔎 Looking in bucket 2 85deb9141ef1eaf725ea4ffb6c4278e75d5a0bdec38e655ecd1ff394441afc43
🔎 Looking in bucket 3 0466a1292a7272e84bc8c1dd6c92de0d413ad5115d27bef60524d250cc627c94
🔎 Looking in bucket 4 474f7af3cebcdb07de10357bb2fe21f7a726b4676a07b1927725ce0aaf263b45
🔎 Looking in bucket 5 cce39f76191873624b6c3cacef7c0ea84aa28cc61eed91837c3008515c8ea1d5
🔎 Looking in bucket 6 14de1cf5ddfa70ef8f193eff96de005650b335303199eceb489627ff7f1998ab
🔎 Looking in bucket 7 d374770c4c495225e235ac89d90586d58910bcf9c5e5386038b950d82294b796
🔎 Looking in bucket 8 7cf48505829612746c83b40c74a79fb846456eed78bc3744f9728859a9660d5b
🔎 Looking in bucket 9 344f47a8f98aca94d7fc37671e676bd4d519bc6dd9aba9214edfe1ed86d39649
🔎 Looking in bucket 10 fe44851e5f29e6a657e76479c46976f2e73c953dc899bd6d3f2b68b94c5b20ca
ℹ️  Found 1 entries
🔎 Looking in bucket 11 8d0a85ef79d8f26b69d2ef539a2fd0cfe8438990f38ae35e73c1f27f943177f2
🔎 Looking in bucket 12 07bac08bd614a20ea079e888b0bfed0bfae7a38bbcefb1b5b09135f791cb006f
🔎 Looking in bucket 13 c258315f26e489954556bd61b0cae4f894d982f63130bfc3697d879165b1ccfd
🔎 Looking in bucket 14 0865c12900fa3652d22a5054d90ef5a346221b773ab4a92c53132a3725515b92
🔎 Looking in bucket 15 4ce19d93f2553f39648347cd1ccc783be9e1ef1d56c7138e68ec7aa7abd67414
🔎 Looking in bucket 16 dc4391bcab88ed82c1421dbf8efa39417b0fe7637b407946cf04e9591db23a2d
🔎 Looking in bucket 17 02f26e79c7a69bb88a6f843c934ba52996c6ea6ae4b9efc5bc5112ba8c77abbc
🔎 Looking in bucket 18 a146de34263062d255a2373c2a0152721bf7e5dece0f12106b065d674c7e2c3b
🔎 Looking in bucket 19 7d678a9925dda72aa91855377a27171472a9e42f22d6bb54b41e45cf3e9ffe88
🔎 Looking in bucket 20 402dba735f087533bc8f574498356ea8b82c56f6d0802b08bbb84e18f452cabc
ℹ️  Adding wasm 438a5528cff17ede6fe515f095c43c5f15727af17d006971485e52462e7e7b89 to search
ℹ️  Found 2 entries
ℹ️  Searching for 0 accounts, 0 contracts, 1 wasms
🔎 Looking in bucket 0 5b071ae54a26e2295e7138c862d01ace6b743ce3acc9e7d6b2b1195426111474
🔎 Looking in bucket 1 503e82f7895a420262dc8ef5e386b1e1cb3155ddee82179ea25567821a0dd8f8
🔎 Looking in bucket 2 85deb9141ef1eaf725ea4ffb6c4278e75d5a0bdec38e655ecd1ff394441afc43
🔎 Looking in bucket 3 0466a1292a7272e84bc8c1dd6c92de0d413ad5115d27bef60524d250cc627c94
🔎 Looking in bucket 4 474f7af3cebcdb07de10357bb2fe21f7a726b4676a07b1927725ce0aaf263b45
🔎 Looking in bucket 5 cce39f76191873624b6c3cacef7c0ea84aa28cc61eed91837c3008515c8ea1d5
🔎 Looking in bucket 6 14de1cf5ddfa70ef8f193eff96de005650b335303199eceb489627ff7f1998ab
🔎 Looking in bucket 7 d374770c4c495225e235ac89d90586d58910bcf9c5e5386038b950d82294b796
🔎 Looking in bucket 8 7cf48505829612746c83b40c74a79fb846456eed78bc3744f9728859a9660d5b
🔎 Looking in bucket 9 344f47a8f98aca94d7fc37671e676bd4d519bc6dd9aba9214edfe1ed86d39649
🔎 Looking in bucket 10 fe44851e5f29e6a657e76479c46976f2e73c953dc899bd6d3f2b68b94c5b20ca
🔎 Looking in bucket 11 8d0a85ef79d8f26b69d2ef539a2fd0cfe8438990f38ae35e73c1f27f943177f2
🔎 Looking in bucket 12 07bac08bd614a20ea079e888b0bfed0bfae7a38bbcefb1b5b09135f791cb006f
🔎 Looking in bucket 13 c258315f26e489954556bd61b0cae4f894d982f63130bfc3697d879165b1ccfd
🔎 Looking in bucket 14 0865c12900fa3652d22a5054d90ef5a346221b773ab4a92c53132a3725515b92
🔎 Looking in bucket 15 4ce19d93f2553f39648347cd1ccc783be9e1ef1d56c7138e68ec7aa7abd67414
🔎 Looking in bucket 16 dc4391bcab88ed82c1421dbf8efa39417b0fe7637b407946cf04e9591db23a2d
🔎 Looking in bucket 17 02f26e79c7a69bb88a6f843c934ba52996c6ea6ae4b9efc5bc5112ba8c77abbc
🔎 Looking in bucket 18 a146de34263062d255a2373c2a0152721bf7e5dece0f12106b065d674c7e2c3b
ℹ️  Found 1 entries
🔎 Looking in bucket 19 7d678a9925dda72aa91855377a27171472a9e42f22d6bb54b41e45cf3e9ffe88
🔎 Looking in bucket 20 402dba735f087533bc8f574498356ea8b82c56f6d0802b08bbb84e18f452cabc
💾 Saved 4 entries to "snapshot.json"
✅ Completed in 3m 9s
```

**Comment by leighmcculloch** [comment]
Todo:
- [ ] automatically pickup trust lines and accounts that hold the asset of a stellar asset contract

File: Makefile [hunk 1]
```diff
@@ -27,7 +27,7 @@ endif
 install_rust: install
 
 install:
-	cargo install --force --locked --path ./cmd/stellar-cli --debug
+	cargo install --force --locked --path ./cmd/stellar-cli #--debug
```
Comment (fnando) [inline Makefile:30]: do we want this?

Comment (leighmcculloch) [reply Makefile:30]: we do not
```suggestion
	cargo install --force --locked --path ./cmd/stellar-cli --debug
```

**Review by fnando — APPROVED** [review]

**Pull request merged**